File: internal/chat/chat_engine.go
package chat

import (
	"archivist/internal/analyzer"
	"archivist/internal/rag"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/redis/go-redis/v9"
)

const (
	// ChatHistoryPrefix is the Redis key prefix for chat histories
	ChatHistoryPrefix = "archivist:chat:history:"
	// ChatHistoryTTL is the TTL for chat histories (24 hours)
	ChatHistoryTTL = 24 * time.Hour
)

// Message represents a single message in a conversation
type Message struct {
	Role      string    `json:"role"`       // "user" or "assistant"
	Content   string    `json:"content"`
	Timestamp time.Time `json:"timestamp"`
	Citations []string  `json:"citations"`  // Source citations for assistant messages
}

// ChatSession represents an ongoing chat session
type ChatSession struct {
	ID            string    `json:"id"`
	PaperTitles   []string  `json:"paper_titles"`
	Messages      []Message `json:"messages"`
	CreatedAt     time.Time `json:"created_at"`
	LastUpdated   time.Time `json:"last_updated"`
}

// ChatEngine handles RAG-powered chat interactions
type ChatEngine struct {
	retriever    *rag.Retriever
	geminiClient *analyzer.GeminiClient
	redisClient  *redis.Client
}

// NewChatEngine creates a new chat engine
func NewChatEngine(retriever *rag.Retriever, geminiClient *analyzer.GeminiClient, redisClient *redis.Client) *ChatEngine {
	return &ChatEngine{
		retriever:    retriever,
		geminiClient: geminiClient,
		redisClient:  redisClient,
	}
}

// StartSession starts a new chat session
func (ce *ChatEngine) StartSession(ctx context.Context, paperTitles []string) (*ChatSession, error) {
	sessionID := fmt.Sprintf("session_%d", time.Now().UnixNano())

	session := &ChatSession{
		ID:          sessionID,
		PaperTitles: paperTitles,
		Messages:    []Message{},
		CreatedAt:   time.Now(),
		LastUpdated: time.Now(),
	}

	// Save to Redis
	if err := ce.saveSession(ctx, session); err != nil {
		return nil, fmt.Errorf("failed to save session: %w", err)
	}

	log.Printf("✓ Started chat session %s with %d papers", sessionID, len(paperTitles))

	return session, nil
}

// Chat processes a user message and generates a response
func (ce *ChatEngine) Chat(ctx context.Context, session *ChatSession, userMessage string) (*Message, error) {
	if userMessage == "" {
		return nil, fmt.Errorf("empty message")
	}

	// Add user message to session
	userMsg := Message{
		Role:      "user",
		Content:   userMessage,
		Timestamp: time.Now(),
	}
	session.Messages = append(session.Messages, userMsg)

	log.Printf("  💬 User: %s", truncateString(userMessage, 60))

	// Retrieve relevant context
	log.Println("  🔍 Retrieving relevant context...")
	var retrievedContext *rag.RetrievedContext
	var err error

	if len(session.PaperTitles) == 0 {
		// Search across all papers
		retrievedContext, err = ce.retriever.Retrieve(ctx, userMessage, nil)
	} else if len(session.PaperTitles) == 1 {
		// Single paper context
		retrievedContext, err = ce.retriever.RetrieveFromPaper(ctx, userMessage, session.PaperTitles[0])
	} else {
		// Multi-paper context
		retrievedContext, err = ce.retriever.RetrieveMultiPaper(ctx, userMessage, session.PaperTitles)
	}

	if err != nil {
		return nil, fmt.Errorf("failed to retrieve context: %w", err)
	}

	log.Printf("  ✓ Retrieved %d relevant chunks", len(retrievedContext.Chunks))

	// Build prompt with context and conversation history
	prompt := ce.buildPrompt(session, userMessage, retrievedContext)

	// Generate response using Gemini
	log.Println("  🤖 Generating response...")
	response, err := ce.geminiClient.GenerateText(ctx, prompt)
	if err != nil {
		return nil, fmt.Errorf("failed to generate response: %w", err)
	}

	// Extract citations
	citations := ce.extractCitations(retrievedContext)

	// Create assistant message
	assistantMsg := Message{
		Role:      "assistant",
		Content:   response,
		Timestamp: time.Now(),
		Citations: citations,
	}

	// Add to session
	session.Messages = append(session.Messages, assistantMsg)
	session.LastUpdated = time.Now()

	// Save updated session
	if err := ce.saveSession(ctx, session); err != nil {
		log.Printf("Warning: failed to save session: %v", err)
	}

	log.Printf("  ✓ Response generated (%d chars)", len(response))

	return &assistantMsg, nil
}

// GetSession retrieves a session from Redis
func (ce *ChatEngine) GetSession(ctx context.Context, sessionID string) (*ChatSession, error) {
	key := ChatHistoryPrefix + sessionID

	data, err := ce.redisClient.Get(ctx, key).Bytes()
	if err == redis.Nil {
		return nil, fmt.Errorf("session not found")
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get session: %w", err)
	}

	var session ChatSession
	if err := json.Unmarshal(data, &session); err != nil {
		return nil, fmt.Errorf("failed to unmarshal session: %w", err)
	}

	return &session, nil
}

// ListSessions lists all active chat sessions
func (ce *ChatEngine) ListSessions(ctx context.Context) ([]*ChatSession, error) {
	pattern := ChatHistoryPrefix + "*"

	keys, err := ce.redisClient.Keys(ctx, pattern).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to list sessions: %w", err)
	}

	sessions := make([]*ChatSession, 0, len(keys))

	for _, key := range keys {
		data, err := ce.redisClient.Get(ctx, key).Bytes()
		if err != nil {
			continue
		}

		var session ChatSession
		if err := json.Unmarshal(data, &session); err != nil {
			continue
		}

		sessions = append(sessions, &session)
	}

	return sessions, nil
}

// DeleteSession deletes a chat session
func (ce *ChatEngine) DeleteSession(ctx context.Context, sessionID string) error {
	key := ChatHistoryPrefix + sessionID

	err := ce.redisClient.Del(ctx, key).Err()
	if err != nil {
		return fmt.Errorf("failed to delete session: %w", err)
	}

	return nil
}

// ExportSessionToLatex exports a chat session to LaTeX format
func (ce *ChatEngine) ExportSessionToLatex(session *ChatSession) string {
	latex := "\\section{Q\\&A Session}\n\n"

	if len(session.PaperTitles) > 0 {
		latex += "\\subsection{Papers Discussed}\n"
		latex += "\\begin{itemize}\n"
		for _, title := range session.PaperTitles {
			latex += fmt.Sprintf("  \\item %s\n", escapeLatex(title))
		}
		latex += "\\end{itemize}\n\n"
	}

	latex += "\\subsection{Conversation}\n\n"

	for i, msg := range session.Messages {
		if msg.Role == "user" {
			latex += fmt.Sprintf("\\textbf{Question %d:} %s\n\n", (i/2)+1, escapeLatex(msg.Content))
		} else {
			latex += fmt.Sprintf("\\textbf{Answer:} %s\n\n", escapeLatex(msg.Content))

			if len(msg.Citations) > 0 {
				latex += "\\textit{Sources:} "
				for j, citation := range msg.Citations {
					if j > 0 {
						latex += ", "
					}
					latex += escapeLatex(citation)
				}
				latex += "\n\n"
			}
		}
	}

	return latex
}

// saveSession saves a session to Redis
func (ce *ChatEngine) saveSession(ctx context.Context, session *ChatSession) error {
	key := ChatHistoryPrefix + session.ID

	data, err := json.Marshal(session)
	if err != nil {
		return fmt.Errorf("failed to marshal session: %w", err)
	}

	err = ce.redisClient.Set(ctx, key, data, ChatHistoryTTL).Err()
	if err != nil {
		return fmt.Errorf("failed to save session to Redis: %w", err)
	}

	return nil
}

// buildPrompt builds the RAG prompt with context and history
func (ce *ChatEngine) buildPrompt(session *ChatSession, userMessage string, context *rag.RetrievedContext) string {
	prompt := "You are a helpful AI research assistant for CS students studying AI/ML papers.\n\n"

	// Add paper context
	if len(session.PaperTitles) > 0 {
		prompt += "You are discussing the following papers:\n"
		for _, title := range session.PaperTitles {
			prompt += fmt.Sprintf("- %s\n", title)
		}
		prompt += "\n"
	}

	// Add retrieved context
	prompt += "RELEVANT CONTEXT FROM PAPERS:\n"
	prompt += "---\n"
	prompt += context.Context
	prompt += "---\n\n"

	// Add conversation history (last 3 exchanges to keep context manageable)
	if len(session.Messages) > 1 {
		prompt += "CONVERSATION HISTORY:\n"
		startIdx := len(session.Messages) - 6 // Last 3 Q&A pairs
		if startIdx < 0 {
			startIdx = 0
		}

		for i := startIdx; i < len(session.Messages)-1; i++ {
			msg := session.Messages[i]
			if msg.Role == "user" {
				prompt += fmt.Sprintf("User: %s\n", msg.Content)
			} else {
				prompt += fmt.Sprintf("Assistant: %s\n", msg.Content)
			}
		}
		prompt += "\n"
	}

	// Add current question
	prompt += "CURRENT QUESTION:\n"
	prompt += userMessage + "\n\n"

	// Add instructions
	prompt += "INSTRUCTIONS:\n"
	prompt += "- Answer the question using the provided context from the papers.\n"
	prompt += "- Be clear, concise, and student-friendly.\n"
	prompt += "- Cite specific sections when referencing information (e.g., 'According to Section 3.2...').\n"
	prompt += "- If the context doesn't contain enough information, say so.\n"
	prompt += "- Use technical terms but explain them when first introduced.\n"
	prompt += "- If comparing multiple papers, clearly distinguish between them.\n\n"

	prompt += "ANSWER:"

	return prompt
}

// extractCitations extracts citation information from retrieved context
func (ce *ChatEngine) extractCitations(context *rag.RetrievedContext) []string {
	citations := []string{}
	seen := make(map[string]bool)

	for _, chunk := range context.Chunks {
		citation := chunk.Document.Source
		if chunk.Document.Section != "" {
			citation += fmt.Sprintf(" (Section: %s)", chunk.Document.Section)
		}

		if !seen[citation] {
			seen[citation] = true
			citations = append(citations, citation)
		}
	}

	return citations
}

// Helper functions

func truncateString(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen] + "..."
}

func escapeLatex(text string) string {
	replacements := map[string]string{
		"\\": "\\textbackslash{}",
		"&":  "\\&",
		"%":  "\\%",
		"$":  "\\$",
		"#":  "\\#",
		"_":  "\\_",
		"{":  "\\{",
		"}":  "\\}",
		"~":  "\\textasciitilde{}",
		"^":  "\\textasciicircum{}",
	}

	for old, new := range replacements {
		text = replaceAll(text, old, new)
	}

	return text
}

func replaceAll(s, old, new string) string {
	result := ""
	for i := 0; i < len(s); i++ {
		found := false
		if i+len(old) <= len(s) && s[i:i+len(old)] == old {
			result += new
			i += len(old) - 1
			found = true
		}
		if !found {
			result += string(s[i])
		}
	}
	return result
}

File: python_rag/chat_engine.py
"""
Chat engine with RAG support
Handles conversations about research papers using retrieval augmented generation
"""
import logging
import time
import json
from typing import List, Optional, Dict, Any
from dataclasses import dataclass, asdict
from pathlib import Path

from .retriever import Retriever, RetrievedContext

logger = logging.getLogger(__name__)


@dataclass
class Message:
    """Chat message"""
    role: str  # "user" or "assistant"
    content: str
    timestamp: float
    citations: List[str] = None

    def __post_init__(self):
        if self.citations is None:
            self.citations = []


@dataclass
class ChatSession:
    """Chat session with history"""
    session_id: str
    paper_titles: List[str]
    messages: List[Message]
    created_at: float
    last_updated: float


class ChatEngine:
    """Chat engine with RAG capabilities"""

    def __init__(
        self,
        retriever: Retriever,
        llm_provider: str = "gemini",
        gemini_api_key: Optional[str] = None,
        model_name: str = "models/gemini-2.0-flash-exp",
        temperature: float = 0.7,
        max_tokens: int = 8000
    ):
        """
        Initialize chat engine

        Args:
            retriever: RAG retriever
            llm_provider: LLM provider ("gemini", "openai")
            gemini_api_key: Gemini API key
            model_name: Model name
            temperature: Sampling temperature
            max_tokens: Maximum tokens to generate
        """
        self.retriever = retriever
        self.llm_provider = llm_provider.lower()
        self.model_name = model_name
        self.temperature = temperature
        self.max_tokens = max_tokens

        # Initialize LLM client
        if self.llm_provider == "gemini":
            if not gemini_api_key:
                raise ValueError("Gemini API key required")

            import google.generativeai as genai
            genai.configure(api_key=gemini_api_key)
            self.model = genai.GenerativeModel(model_name)
            logger.info(f"✓ Gemini chat engine initialized: {model_name}")

        elif self.llm_provider == "openai":
            if not gemini_api_key:  # Reuse for OpenAI key
                raise ValueError("OpenAI API key required")

            from openai import OpenAI
            self.client = OpenAI(api_key=gemini_api_key)
            logger.info(f"✓ OpenAI chat engine initialized: {model_name}")

        else:
            raise ValueError(f"Unsupported LLM provider: {llm_provider}")

        # Session storage
        self.sessions: Dict[str, ChatSession] = {}

    def create_session(self, paper_titles: List[str]) -> ChatSession:
        """
        Create a new chat session

        Args:
            paper_titles: Papers to chat about

        Returns:
            ChatSession
        """
        session_id = f"session_{int(time.time() * 1000)}"
        now = time.time()

        session = ChatSession(
            session_id=session_id,
            paper_titles=paper_titles,
            messages=[],
            created_at=now,
            last_updated=now
        )

        self.sessions[session_id] = session

        logger.info(f"✓ Created chat session {session_id} with {len(paper_titles)} papers")

        return session

    def chat(
        self,
        session_id: str,
        user_message: str,
        stream: bool = False
    ) -> Message:
        """
        Process a chat message with RAG

        Args:
            session_id: Session ID
            user_message: User's message
            stream: Whether to stream response (not implemented)

        Returns:
            Assistant message with response and citations
        """
        if session_id not in self.sessions:
            raise ValueError(f"Session not found: {session_id}")

        session = self.sessions[session_id]

        logger.info(f"💬 User: {self._truncate(user_message, 60)}")

        # Add user message to history
        user_msg = Message(
            role="user",
            content=user_message,
            timestamp=time.time()
        )
        session.messages.append(user_msg)

        # Retrieve relevant context
        logger.info("  🔍 Retrieving relevant context...")

        if not session.paper_titles:
            # No specific papers, search all
            context = self.retriever.retrieve(user_message)
        elif len(session.paper_titles) == 1:
            # Single paper
            context = self.retriever.retrieve_from_paper(
                user_message,
                session.paper_titles[0]
            )
        else:
            # Multiple papers
            context = self.retriever.retrieve_multi_paper(
                user_message,
                session.paper_titles
            )

        logger.info(f"  ✓ Retrieved {context.total_chunks} chunks from {len(context.sources)} sources")

        # Build prompt
        prompt = self._build_prompt(session, user_message, context)

        # Generate response
        logger.info("  🤖 Generating response...")
        response_text = self._generate_response(prompt)

        # Extract citations
        citations = self._extract_citations(context)

        # Create assistant message
        assistant_msg = Message(
            role="assistant",
            content=response_text,
            timestamp=time.time(),
            citations=citations
        )

        # Add to session
        session.messages.append(assistant_msg)
        session.last_updated = time.time()

        logger.info(f"  ✓ Response generated ({len(response_text)} chars)")

        return assistant_msg

    def get_session(self, session_id: str) -> Optional[ChatSession]:
        """Get a chat session"""
        return self.sessions.get(session_id)

    def get_all_sessions(self) -> List[ChatSession]:
        """Get all active sessions"""
        return list(self.sessions.values())

    def delete_session(self, session_id: str) -> bool:
        """Delete a session"""
        if session_id in self.sessions:
            del self.sessions[session_id]
            logger.info(f"✓ Deleted session {session_id}")
            return True
        return False

    def export_session_to_latex(self, session_id: str) -> str:
        """
        Export chat session to LaTeX format

        Args:
            session_id: Session ID

        Returns:
            LaTeX formatted conversation
        """
        if session_id not in self.sessions:
            raise ValueError(f"Session not found: {session_id}")

        session = self.sessions[session_id]

        latex = "\\section{Q\\&A Session}\n\n"

        if session.paper_titles:
            latex += "\\subsection{Papers Discussed}\n"
            latex += "\\begin{itemize}\n"
            for title in session.paper_titles:
                latex += f"  \\item {self._escape_latex(title)}\n"
            latex += "\\end{itemize}\n\n"

        latex += "\\subsection{Conversation}\n\n"

        qa_number = 1
        for msg in session.messages:
            if msg.role == "user":
                latex += f"\\textbf{{Question {qa_number}:}} {self._escape_latex(msg.content)}\n\n"
            else:
                latex += f"\\textbf{{Answer:}} {self._escape_latex(msg.content)}\n\n"

                if msg.citations:
                    latex += "\\textit{Sources:} "
                    latex += ", ".join(self._escape_latex(c) for c in msg.citations)
                    latex += "\n\n"

                qa_number += 1

        return latex

    def _build_prompt(
        self,
        session: ChatSession,
        user_message: str,
        context: RetrievedContext
    ) -> str:
        """Build RAG prompt with context and history"""
        prompt_parts = []

        # System instruction
        prompt_parts.append(
            "You are a helpful AI research assistant for CS students studying AI/ML, "
            "Computer Vision, and Networking papers.\n\n"
        )

        # Papers being discussed
        if session.paper_titles:
            prompt_parts.append("You are discussing the following papers:\n")
            for title in session.paper_titles:
                prompt_parts.append(f"- {title}\n")
            prompt_parts.append("\n")

        # Retrieved context
        prompt_parts.append("RELEVANT CONTEXT FROM PAPERS:\n")
        prompt_parts.append("---\n")
        prompt_parts.append(context.context_text)
        prompt_parts.append("---\n\n")

        # Conversation history (last 3 exchanges)
        if len(session.messages) > 1:
            prompt_parts.append("CONVERSATION HISTORY:\n")

            # Get last 6 messages (3 Q&A pairs)
            history_messages = session.messages[max(0, len(session.messages) - 7):-1]

            for msg in history_messages:
                if msg.role == "user":
                    prompt_parts.append(f"User: {msg.content}\n")
                else:
                    prompt_parts.append(f"Assistant: {msg.content}\n")

            prompt_parts.append("\n")

        # Current question
        prompt_parts.append("CURRENT QUESTION:\n")
        prompt_parts.append(f"{user_message}\n\n")

        # Instructions
        prompt_parts.append("INSTRUCTIONS:\n")
        prompt_parts.append("- Answer the question using the provided context from the papers.\n")
        prompt_parts.append("- Be clear, concise, and student-friendly.\n")
        prompt_parts.append("- Cite specific sections when referencing information.\n")
        prompt_parts.append("- If the context doesn't contain enough information, say so.\n")
        prompt_parts.append("- Use technical terms but explain them when first introduced.\n")
        prompt_parts.append("- If comparing multiple papers, clearly distinguish between them.\n\n")

        prompt_parts.append("ANSWER:\n")

        return "".join(prompt_parts)

    def _generate_response(self, prompt: str) -> str:
        """Generate LLM response"""
        if self.llm_provider == "gemini":
            response = self.model.generate_content(
                prompt,
                generation_config={
                    'temperature': self.temperature,
                    'max_output_tokens': self.max_tokens,
                }
            )
            return response.text

        elif self.llm_provider == "openai":
            response = self.client.chat.completions.create(
                model=self.model_name,
                messages=[{"role": "user", "content": prompt}],
                temperature=self.temperature,
                max_tokens=self.max_tokens
            )
            return response.choices[0].message.content

        else:
            raise NotImplementedError(f"LLM provider not implemented: {self.llm_provider}")

    def _extract_citations(self, context: RetrievedContext) -> List[str]:
        """Extract citation strings from retrieved context"""
        citations = []
        seen = set()

        for chunk in context.chunks:
            doc = chunk.document
            source = doc.metadata.get('source', 'unknown')
            section = doc.metadata.get('section', '')

            if section:
                citation = f"{source} (Section: {section})"
            else:
                citation = source

            if citation not in seen:
                seen.add(citation)
                citations.append(citation)

        return citations

    @staticmethod
    def _truncate(text: str, max_len: int) -> str:
        """Truncate text"""
        if len(text) <= max_len:
            return text
        return text[:max_len] + "..."

    @staticmethod
    def _escape_latex(text: str) -> str:
        """Escape LaTeX special characters"""
        replacements = {
            '\\': '\\textbackslash{}',
            '&': '\\&',
            '%': '\\%',
            '$': '\\$',
            '#': '\\#',
            '_': '\\_',
            '{': '\\{',
            '}': '\\}',
            '~': '\\textasciitilde{}',
            '^': '\\textasciicircum{}',
        }

        for char, replacement in replacements.items():
            text = text.replace(char, replacement)

        return text


def create_chat_engine(
    retriever: Retriever,
    gemini_api_key: Optional[str] = None,
    model_name: str = "models/gemini-2.0-flash-exp",
    temperature: float = 0.7
) -> ChatEngine:
    """
    Factory function to create chat engine

    Args:
        retriever: RAG retriever
        gemini_api_key: Gemini API key
        model_name: Model name
        temperature: Sampling temperature

    Returns:
        ChatEngine instance
    """
    return ChatEngine(
        retriever=retriever,
        llm_provider="gemini",
        gemini_api_key=gemini_api_key,
        model_name=model_name,
        temperature=temperature
    )
